`doNotSerialize`, `jsonName` pragmas for JSON serialization closes #8104, #10718, also fixes #11415

Vindaar opened the pull request:
This PR adds the ability to exclude certain object fields from JSON serialization via `%`, by annotating the field with the `{.noserialize.}` pragma.

```nim
type
  User = object
    name: string
    age: int
    uid {.noserialize.}: int
let user = User(name: "Siri", age: 7, uid: 1234)
let uJson = % user
doAssert not uJson.hasKey("uid")
```

After discussion on IRC:
https://irclogs.nim-lang.org/05-06-2019.html#19:15:24
@dom96 asked me to create a PR for this.

While working on it I stumbled upon an issue with custom pragmas on variant objects: #11415.

However, to not break things, I also had to extend how `typDef` is determined in `customPragmaNode` (see https://github.com/nim-lang/Nim/pull/11416/commits/5da931fe811717a45f2dd272ea6281979c3e8f0b). This is a problem for the first test case in `tjsonmacro.nim`:
https://github.com/nim-lang/Nim/blob/devel/tests/stdlib/tjsonmacro.nim#L9-L45
Line 45 will fail with `node is not a symbol`. I couldn't really reproduce this without the `%` proc change, which is why I didn't open an issue for this as well.

#### Update: add `jsonName`

I also added a `jsonName` pragma, to customize the (de)serialization name of an object field. 

```nim
type
  User = object
    name: string
    age {.jsonName: "userAge".}: int
let user = User(name: "Siri", age: 7)
let uJson = % user
doAssert uJson["userAge"].num == 7
```

This pragma closes #10718.

dom96 commented on lib/pure/json.nim:
```diff
@@ -363,10 +392,25 @@ proc `[]=`*(obj: JsonNode, key: string, val: JsonNode) {.inline.} =
+
 proc `%`*[T: object](o: T): JsonNode =
   ## Construct JsonNode from tuples and objects.
+  ## An object field annotated with the `{.noserialize.}` pragma will not appear
```
```suggestion
  ##
  ## An object field annotated with the `{.noserialize.}` pragma will not appear
```

RST rendering will not put a line break unless you include an empty line like I have done above

Vindaar replied:
Oh, thanks! I'm not very experienced with RST.

dom96 commented on lib/pure/json.nim:
```diff
@@ -363,10 +392,25 @@ proc `[]=`*(obj: JsonNode, key: string, val: JsonNode) {.inline.} =
   assert(obj.kind == JObject)
   obj.fields[key] = val
 
+template noserialize*() {.pragma.}
```
If possible please add documentation to this as well.

Vindaar replied:
Will do!

dom96 commented on lib/pure/json.nim:
```diff
@@ -116,6 +116,35 @@
+## .. code-block:: nim
+##   import json
+##   let number = 10
+##   let numJson = % number
```
Nit: no space after ``%``. Please change this below too.

dom96 commented on lib/pure/json.nim:
```diff
@@ -116,6 +116,35 @@
+##
+## .. code-block:: nim
+##   import json
+##   let number = 10
```
I would simplify this and remove this variable, simply use the literal below.

dom96 commented on lib/pure/json.nim:
```diff
@@ -116,6 +116,35 @@
+##   doAssert numJson.kind == JInt
+##
+## For objects it may be desirable to avoid serialization of one or more object
+## fields. This can be achieved by using the ``{.noserialize.}`` pragma.
```
Let's use camelCase here (and everywhere): ``{.noSerialize.}``

Vindaar commented:
Sorry, I overlooked your comment. I'll address the comments later today.

I have no strong feelings about the pragma name. Just picked the first thing that came to mind. Although I might actually prefer the `{.json: WhatEver.}` syntax.

alehander42 commented:
keep in mind i can have a "NoExport" field in my json, and this might make a bit confusing `json: NoExport`: yes, i know it doesnt have `"` but still adds a moment of "let me think about it while i read it

dom96 commented:
@alehander42 got any better suggestions?

Vindaar commented:
Ok, I addressed the comments. For now I only changed the name to camelCase instead of changing it completely. I don't want to be the person making that decision.

disruptek commented:
I like `noSerialize` because it's a notation that transcends implementation.  JSON is but one of many possible serialization formats and by abstracting the concept, neither the code nor the programmer needs to change to adapt to new formats.

hlaaftana commented:
Don't mean to make things more complicated, just adding to the discussion:

If `{.json: "CustomFieldName".}` were to be implemented, you could also implement multiple aliases for the same field, as in `{.json: ["CustomFieldName1", "CustomFieldName2"].}`. You could even support things like `field {.json: ["UnparsedInt": string].}: int` or `field {.json: string.}: int` to convert types during marshalling. This would allow for either `{.json: [].}` or `{.json: void.}` to mean no serialization at all, but it looks kind of ugly.

If you still want to just use a name, you could still do something like `{.json: noserialize.}` and have other libraries do `{.otherformat: noserialize.}` as well. The `transient` keyword in java is also an option for a name as it serves the same purpose.

zah commented:
In our developing generic serialization library, I've used a pragma called `dontSerialize`. That's longer than `noSerialize`, but it's also more proper grammar :P

https://github.com/status-im/nim-serialization/

dom96 commented:
@zah ``doNotSerialize`` would be better :P

Vindaar commented:
I renamed `noSerialize` to `doNotSerialize` and rebased on the current devel.

Vindaar commented:
Due to @Phillips126 comment in #10718, I saw Araq's comment about adding a `jsonName` pragma to customize the (de)serialization names.

Since this PR is related anyways I  thought I'd attempt to add this as well. Due to the addition of the `jsonName` pragma, the naming of the `doNotSerialize` pragma is probably open for bikeshedding again. :P (or if it should just be the result of `{.jsonName: "".}` instead of having a different pragma for no serialization.

The implementation isn't very pretty unfortunately for the deserialization, because the whole deserialization logic works on the type node obtained via `getTypeImpl`. The custom pragma information however is only available when calling `getImpl` on the type symbol. So I ended up implementing this feature by fixing up the code after the macro is done.

Vindaar commented:
With https://github.com/nim-lang/Nim/pull/12391 now merged, I'll close this PR and rewrite it. Will probably create a new PR once everything is working.

disruptek commented:
Any thoughts about a new version of this PR?

Vindaar commented:
@disruptek Sorry, I kinda forgot about it. 

I'll try to take a stab at it in the next few days!

Vindaar commented:
I just remembered why I put this on ice. It's because I was waiting for https://github.com/nim-lang/Nim/pull/11526 to be merged.

I just started to reimplement this. `doNotSerialize` is essentially the same as in this PR. But the `jsonName` part doesn't seem to be possible right now (without making some changes to the `to` procedures), because in the current logic I just can't seem to find a way to extract the pragma information from the types in `foldObjectBody`.
I'd have to extract the type information from the `typedesc` in `to` directly. This would then to require to pass that information to `initFromJson`, which is ugly. Would have to change the signature either of all `initFromJson` procs (which is bad, because only one needs the replacement information) or only for the `object | tuple` proc, but then I'd have to make sure to call the correct proc within the macros depending on the symbols type. Makes for ugly code.

So, I'll wait for a bit until the above PR is hopefully merged soon.

brentp commented:
bump. this would be quite useful.

ghost commented:
@Vindaar Well yes, you need to modify the .to procedure, you can check my custom implementation which is based on that PR - https://github.com/Yardanico/telenim/blob/master/src/telenim/json_custom.nim (called toCustom here)

Vindaar commented:
@Yardanico Uhm, if you already have a working version of this around, feel free to create a PR for this!
If you don't feel like doing that, I'll take a look at it. I'm still not sure though about the custom pragmas implementation / would rather wait until that PR is merged.

ghost commented:
@Vindaar yeah, we should want until #11526 is merged because I just copied needed parts of it into my json_custom

Clyybber commented:
https://github.com/nim-lang/Nim/pull/11526 is now merged :)
